## 1. Analyzing SAT Score Data of Various Schools
<p>Analyszed in this project are SAT scores for reading, math, and writing, each with a maximum score of 800 points.

<p>As my family is currently running a small pre-school in the US, I wanted to research some data of the development for the youth in a certain region, in this case in New York.

Please do go checkout the Website i developed for our pre-school if interested at: https://kidspalace.org/


 For a wide range of stakeholders, including policy and education experts, researchers, the government, and even parents deciding which school their children should attend for their development, examining the performance of schools is vital. </p>



<p>In this notebook, we will take a look at data on SATs across public schools in New York City. Our database contains a single table:</p>

<h1 id="schools"><code>schools</code></h1>
<table>
<thead>
<tr>
<th>column</th>
<th>type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>school_name</code></td>
<td><code>varchar</code></td>
<td>Name of school</td>
</tr>
<tr>
<td><code>borough</code></td>
<td><code>varchar</code></td>
<td>Borough that the school is located in</td>
</tr>
<tr>
<td><code>building_code</code></td>
<td><code>varchar</code></td>
<td>Code for the building</td>
</tr>
<tr>
<td><code>average_math</code></td>
<td><code>int</code></td>
<td>Average math score for SATs</td>
</tr>
<tr>
<td><code>average_reading</code></td>
<td><code>int</code></td>
<td>Average reading score for SATs</td>
</tr>
<tr>
<td><code>average_writing</code></td>
<td><code>int</code></td>
<td>Average writing score for SATs</td>
</tr>
<tr>
<td><code>percent_tested</code></td>
<td><code>numeric</code></td>
<td>Percentage of students completing SATs</td>
</tr>
</tbody>
</table>
<p>Let's familiarize ourselves with the data by taking a looking at the first few schools!</p>

In [1]:
import sqlite3
import pandas as pd
import os
import numpy as np
import my_model.metrics.precision
# or
from my_model.metrics import precision


In [2]:
!pip install ipython-sql

  Using cached ipython_sql-0.4.1-py3-none-any.whl (21 kB)
  Using cached sqlparse-0.4.3-py3-none-any.whl (42 kB)
     ---------------------------------------- 1.6/1.6 MB 279.7 kB/s eta 0:00:00
  Using cached prettytable-0.7.2.zip (28 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
     ------------------------------------ 190.9/190.9 kB 361.3 kB/s eta 0:00:00
Using legacy 'setup.py install' for prettytable, since package 'wheel' is not installed.
  Running setup.py install for prettytable: started
  Running setup.py install for prettytable: finished with status 'done'



[notice] A new release of pip available: 22.2.1 -> 22.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
!pip install sqlite3

ERROR: Could not find a version that satisfies the requirement sqlite3 (from versions: none)
ERROR: No matching distribution found for sqlite3

[notice] A new release of pip available: 22.2.1 -> 22.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
df = pd.read_csv(r"C:\Users\smukh\Documents\Data Analysis Projects\PROJECT-NYC_Public_School_Test_Analysis-main\dataset\schools_modified.csv")
df



school_name    borough  \
0    New Explorations into Science, Technology and ...  Manhattan   
1                                 Essex Street Academy  Manhattan   
2                         Lower Manhattan Arts Academy  Manhattan   
3      High School for Dual Language and Asian Studies  Manhattan   
4        Henry Street School for International Studies  Manhattan   
..                                                 ...        ...   
370  Queens High School for Information, Research, ...     Queens   
371  Rockaway Park High School for Environmental Su...     Queens   
372                   Channel View School for Research     Queens   
373                    Rockaway Collegiate High School     Queens   
374                                  Scholars' Academy     Queens   

    building_code  average_math  average_reading  average_writing  \
0            M022           657              601              601   
1            M445           395              411              387   
2            M445           418              428              415   
3            M445           613              453              463   
4            M056           410              406              381   
..            ...           ...              ...              ...   
370          Q465           372              362              352   
371          Q410           357              381              376   
372          Q410           427              430              423   
373          Q410           399              403              405   
374          Q180           588              560              568   

     percent_tested  
0               NaN  
1              78.9  
2              65.1  
3              95.9  
4              59.7  
..              ...  
370            44.6  
371            38.5  
372            76.6  
373            46.5  
374            99.2  

[375 rows x 7 columns]

In [6]:
df.describe()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


In [7]:
cnn = sqlite3.connect('jupyter_sql_tutorial.db')


In [8]:
df.to_sql('people',cnn)


375

In [10]:
%load_ext sql

In [17]:
%sql sqlite:/// jupyter_sql_tutorial.db 


%sql SELECT * FROM people LIMIT 5;


(sqlite3.OperationalError) near "jupyter_sql_tutorial": syntax error
[SQL: jupyter_sql_tutorial.db #is this the right way to connect to the database??]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
 * sqlite:///
(sqlite3.OperationalError) no such table: people
[SQL: SELECT * FROM people LIMIT 5;]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [ ]:
%%sql

In [15]:
%%sql
postgresql:///schools
    
-- Select all columns from the database
-- Display only the first ten rows

SELECT *
FROM schools
LIMIT 10

Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys(['sqlite:///'])
No module named 'psycopg2'
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys(['sqlite:///'])


## 2. Finding missing values
<p>It looks like the first school in our database had no data in the <code>percent_tested</code> column! </p>
<p>Let's identify how many schools have missing data for this column, indicating schools that did not report the percentage of students tested. </p>
<p>To understand whether this missing data problem is widespread in New York, we will also calculate the total number of schools in the database.</p>

In [16]:
%%sql

-- Count rows with percent_tested missing and total number of schools

SELECT COUNT(*)-COUNT(percent_tested) num_tested_missing, count(*) num_schools
FROM schools


 * sqlite:///
(sqlite3.OperationalError) no such table: schools
[SQL: -- Count rows with percent_tested missing and total number of schools

SELECT COUNT(*)-COUNT(percent_tested) num_tested_missing, count(*) num_schools
FROM schools]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


## 3. Schools by building code
<p>There are 20 schools with missing data for <code>percent_tested</code>, which only makes up 5% of all rows in the database.</p>
<p>Now let's turn our attention to how many schools there are. When we displayed the first ten rows of the database, several had the same value in the <code>building_code</code> column, suggesting there are multiple schools based in the same location. Let's find out how many unique school locations exist in our database. </p>

In [196]:
%%sql

-- Count the number of unique building_code values

SELECT COUNT(DISTINCT building_code) num_school_buildings
FROM schools

 * postgresql:///schools
1 rows affected.


[(233,)]

## 4. Best schools for math
<p>Out of 375 schools, only 233 (62%) have a unique <code>building_code</code>! </p>
<p>Now let's start our analysis of school performance. As each school reports individually, we will treat them this way rather than grouping them by <code>building_code</code>. </p>
<p>First, let's find all schools with an average math score of at least 80% (out of 800). </p>

In [198]:
%%sql

-- Select school and average_math
-- Filter for average_math 640 or higher
-- Display from largest to smallest average_math

SELECT school_name, average_math
FROM schools
WHERE average_math >= 800*0.8
ORDER BY average_math DESC

 * postgresql:///schools
10 rows affected.


[('Stuyvesant High School', 754),
 ('Bronx High School of Science', 714),
 ('Staten Island Technical High School', 711),
 ('Queens High School for the Sciences at York College', 701),
 ('High School for Mathematics, Science, and Engineering at City College', 683),
 ('Brooklyn Technical High School', 682),
 ('Townsend Harris High School', 680),
 ('High School of American Studies at Lehman College', 669),
 ('New Explorations into Science, Technology and Math High School', 657),
 ('Eleanor Roosevelt High School', 641)]

## 5. Lowest reading score
<p>Wow, there are only ten public schools in New York City with an average math score of at least 640!</p>
<p>Now let's look at the other end of the spectrum and find the single lowest score for reading. We will only select the score, not the school, to avoid naming and shaming!</p>

In [200]:
%%sql

-- Find lowest average_reading

SELECT min(average_reading) lowest_reading
FROM schools

 * postgresql:///schools
1 rows affected.


[(302,)]

## 6. Best writing school
<p>The lowest average score for reading across schools in New York City is less than 40% of the total available points! </p>
<p>Now let's find the school with the highest average writing score.</p>

In [202]:
%%sql

-- Find the top score for average_writing
-- Group the results by school
-- Sort by max_writing in descending order
-- Reduce output to one school

SELECT school_name, average_writing max_writing
FROM schools
ORDER BY max_writing DESC
LIMIT 1

 * postgresql:///schools
1 rows affected.


[('Stuyvesant High School', 693)]

## 7. Top 10 schools
<p>An average writing score of 693 is pretty impressive! </p>
<p>This top writing score was at the same school that got the top math score, Stuyvesant High School. Stuyvesant is widely known as a perennial top school in New York. </p>
<p>What other schools are also excellent across the board? Let's look at scores across reading, writing, and math to find out.</p>

In [204]:
%%sql

-- Calculate average_sat
-- Group by school_name
-- Sort by average_sat in descending order
-- Display the top ten results

SELECT school_name, sum(average_math + average_reading + average_writing) average_sat
FROM schools
GROUP BY school_name
ORDER BY average_sat DESC
LIMIT 10

 * postgresql:///schools
10 rows affected.


[('Stuyvesant High School', 2144),
 ('Staten Island Technical High School', 2041),
 ('Bronx High School of Science', 2041),
 ('High School of American Studies at Lehman College', 2013),
 ('Townsend Harris High School', 1981),
 ('Queens High School for the Sciences at York College', 1947),
 ('Bard High School Early College', 1914),
 ('Brooklyn Technical High School', 1896),
 ('Eleanor Roosevelt High School', 1889),
 ('High School for Mathematics, Science, and Engineering at City College', 1889)]

## 8. Ranking boroughs
<p>There are four schools with average SAT scores of over 2000! Now let's analyze performance by New York City borough. </p>
<p>We will build a query that calculates the number of schools and the average SAT score per borough!</p>

In [206]:
%%sql

-- Select borough and a count of all schools, aliased as num_schools
-- Calculate the sum of average_math, average_reading, and average_writing, divided by a count of all schools, aliased as average_borough_sat
-- Organize results by borough
-- Display by average_borough_sat in descending order

SELECT borough, count(*) num_schools, sum(average_math + average_reading + average_writing)/count(*) average_borough_sat
FROM schools
GROUP BY borough
ORDER BY average_borough_sat DESC


 * postgresql:///schools
5 rows affected.


[('Staten Island', 10, 1439),
 ('Queens', 69, 1345),
 ('Manhattan', 89, 1340),
 ('Brooklyn', 109, 1230),
 ('Bronx', 98, 1202)]

## 9. Brooklyn numbers
<p>It appears that schools in Staten Island, on average, produce higher scores across all three categories. However, there are only 10 schools in Staten Island, compared to an average of 91 schools in the other four boroughs!</p>
<p>For our final query of the database, let's focus on Brooklyn, which has 109 schools. We wish to find the top five schools for math performance.</p>

In [208]:
%%sql

-- Select school and average_math
-- Filter for schools in Brooklyn
-- Aggregate on school_name
-- Display results from highest average_math and restrict output to five rows

SELECT school_name, average_math
FROM schools
WHERE borough = 'Brooklyn'
ORDER BY average_math DESC
LIMIT 5

 * postgresql:///schools
5 rows affected.


[('Brooklyn Technical High School', 682),
 ('Brooklyn Latin School', 625),
 ('Leon M. Goldstein High School for the Sciences', 563),
 ('Millennium Brooklyn High School', 553),
 ('Midwood High School', 550)]